pennylane/ops/op_math/prod.py

```diff
@@ -0,0 +1,273 @@
+"""
+This file contains the implementation of the Prod class which contains logic for
+computing the product between operations.
+"""
+from copy import copy
+from functools import reduce
+from itertools import combinations
+
+import numpy as np
+
+import pennylane as qml
+from pennylane import math
+from pennylane.operation import Operator, expand_matrix
+
+
+def prod(*ops, do_queue=True, id=None):
+    """Construct an operator which represents the generalized product of the
+    operators provided.
+
+    The generalized product operation represents both the tensor product as
+    well as matrix composition. This can be resolved naturally from the wires
+    that the given operators act on.
+
+    Args:
+        ops (tuple[~.operation.Operator]): The operators we would like to multiply
+
+    Keyword Args:
+        do_queue (bool): determines if the product operator will be queued
+            (currently not supported). Default is True.
+        id (str or None): id for the product operator. Default is None.
+
+    Returns:
+        ~ops.op_math.Prod: the operator representing the product.
+
+    ..seealso:: :class:`~.ops.op_math.Prod`
+
+    **Example**
+
+    >>> prod_op = prod(qml.PauliX(0), qml.PauliZ(0))
+    >>> prod_op
+    PauliX(wires=[0]) @ PauliZ(wires=[0])
+    >>> prod_op.matrix()
+    array([[ 0, -1],
+           [ 1,  0]])
+    """
+    return Prod(*ops, do_queue=do_queue, id=id)
+
+
+class Prod(Operator):
+    r"""Symbolic operator representing the product of operators.
+
+    Args:
+        factors (tuple[~.operation.Operator]): a tuple of operators which will be multiplied together.
+
+    Keyword Args:
+        do_queue (bool): determines if the product operator will be queued (currently not supported).
+            Default is True.
+        id (str or None): id for the product operator. Default is None.
+
+    **Example**
+
+    >>> prop_op = Prod(qml.PauliX(0), qml.PauliZ(0))
+    >>> prop_op
+    PauliX(wires=[0]) @ PauliZ(wires=[0])
+    >>> qml.matrix(prop_op)
+    array([[ 0,  -1],
+           [ 1,   0]])
+    >>> prop_op.terms()
+    ([1.0], [PauliX(wires=[0]) @ PauliZ(wires=[0])])
+
+    .. details::
+        :title: Usage Details
+
+        The Prod operator represents both matrix composition and tensor products
+        between operators.
+
+        >>> prod_op = Prod(qml.RZ(1.23, wires=0), qml.PauliX(wires=0), qml.PauliZ(wires=1))
+        >>> prod_op.matrix()
+        array([[ 0.        +0.j        ,  0.        +0.j        ,
+                 0.81677345-0.57695852j,  0.        +0.j        ],
+               [ 0.        +0.j        ,  0.        +0.j        ,
+                 0.        +0.j        , -0.81677345+0.57695852j],
+               [ 0.81677345+0.57695852j,  0.        +0.j        ,
+                 0.        +0.j        ,  0.        +0.j        ],
+               [ 0.        +0.j        , -0.81677345-0.57695852j,
+                 0.        +0.j        ,  0.        +0.j        ]])
+    """
```

Comment:
Should note that when used in a circuit, `Prod(op1, op2)` indicates first applying `op2` and then `op1`.  We can show this in an example via `op.decompostion()`.

Reply:
Added 👍🏼

Reply:
Thanks 👍